File: PriorityQ.java
import java.util.*;
import java.util.stream.Collectors;
import java.util.stream.Stream;
import java.io.*;

public class PriorityQ {
    public static void main(String[] args) throws IOException {
        PriorityQueue<List<Integer>> pQ = new PriorityQueue(new Comparator<List<Integer>>() {
            public int compare(List<Integer> left, List<Integer> right) {
                if (left.get(1) == right.get(1)) {
                    return left.get(0) - right.get(0);
                }
                return left.get(1) - right.get(1);
            }
        });

        int[][] temp = {{3, 6, 4}, {4, 2, 5}, {1, 0, 5}, {5, 0, 5}};
        for (int i = 0; i < temp.length; ++i) {
            ArrayList<Integer> lst = new ArrayList();
            for (int j = 0; j < 3; ++j) {
                lst.add(temp[i][j]);
            }
            pQ.add(lst);
        }

        //0: priorityScore, 1: startTime, 2: execTime
        long[] answer = new long[11];
        while (!pQ.isEmpty()) {
            List<Integer> curr = pQ.poll();

            System.out.println(curr);
            answer[curr.get(0)] += answer[0] - curr.get(1);
            answer[0] += curr.get(2);
        }
        for (int i = 0; i < 11; ++i) {
            System.out.print(answer[i] + " ");
        }
        System.out.println();
    }
}

File: PriorityQ.kt
import java.util.PriorityQueue

fun main() {
    val pq = PriorityQueue<Int>()
    val pq2 = PriorityQueue { p1: Pair<Int, Int>, p2: Pair<Int, Int> ->
        if (p1.second > p2.second) 1 else if (p1.second == p2.second) 0 else -1
    }

    pq += 5
    pq += 1
    pq += 3

//    pq2 += Pair(1,4)
//    pq2 += Pair(2,3)
//    pq2 += Pair(3,2)
    while (true) {
        println(pq2.peek())
    }
    /* remove = 실패 시, NoSuchElementException 발생
     * poll = 실패 시, null 반환
     * add = 실패 시, IllegalStateException 발생
     * offer = 실패 시, false 반환
     * peek = 실패 시, null 반환
     */
}